File: axon-sdk-python/axon/messages.py
from axon._base import _BaseClient, _BaseSyncClient

class MessagesClient(_BaseClient):
    async def send(
        self,
        recipient_id: str = None,
        topic: str = None,
        payload: dict = None,
    ) -> dict:
        """
        Send a direct message or publish to a topic (async).
        """
        body = {}
        if recipient_id is not None:
            body["recipient_id"] = recipient_id
        if topic is not None:
            body["topic"] = topic
        if payload is not None:
            body["payload"] = payload

        return await self._request("POST", "/v1/messages/send", json=body)

    async def get_inbox(self, topic: str = None, limit: int = 50) -> list[dict]:
        """
        Retrieve direct messages or topic updates in the inbox (async).
        """
        params = {"limit": limit}
        if topic is not None:
            params["topic"] = topic
        data = await self._request("GET", "/v1/messages/inbox", params=params)
        return data.get("messages", [])

    async def ack(self, message_id: str) -> bool:
        """
        Acknowledge a received message (async).
        """
        data = await self._request(
            "POST",
            "/v1/messages/ack",
            params={"message_id": message_id}
        )
        return data.get("acknowledged", False)


class SyncMessagesClient(_BaseSyncClient):
    def send(
        self,
        recipient_id: str = None,
        topic: str = None,
        payload: dict = None,
    ) -> dict:
        """
        Send a direct message or publish to a topic (sync).
        """
        body = {}
        if recipient_id is not None:
            body["recipient_id"] = recipient_id
        if topic is not None:
            body["topic"] = topic
        if payload is not None:
            body["payload"] = payload

        return self._request("POST", "/v1/messages/send", json=body)

    def get_inbox(self, topic: str = None, limit: int = 50) -> list[dict]:
        """
        Retrieve direct messages or topic updates in the inbox (sync).
        """
        params = {"limit": limit}
        if topic is not None:
            params["topic"] = topic
        data = self._request("GET", "/v1/messages/inbox", params=params)
        return data.get("messages", [])

    def ack(self, message_id: str) -> bool:
        """
        Acknowledge a received message (sync).
        """
        data = self._request(
            "POST",
            "/v1/messages/ack",
            params={"message_id": message_id}
        )
        return data.get("acknowledged", False)

File: axon-sdk-js/src/messages.ts
// ─── Axon Protocol SDK — Messages Module ───

import type { HttpClient } from './_base.js';
import type { MessageSendRequest, MessageSendResponse, MessageInfo } from './types.js';

export class MessagesModule {
  constructor(private readonly http: HttpClient) {}

  /**
   * Send a direct message or publish to a topic.
   */
  async send(request: MessageSendRequest): Promise<MessageSendResponse> {
    return this.http.post<MessageSendResponse>('/v1/messages/send', request);
  }

  /**
   * Retrieve direct messages or topic updates in the inbox.
   */
  async inbox(params: { topic?: string; limit?: number } = {}): Promise<MessageInfo[]> {
    const queryParams: Record<string, string> = {};
    if (params.limit !== undefined) {
      queryParams.limit = String(params.limit);
    }
    if (params.topic !== undefined) {
      queryParams.topic = params.topic;
    }
    const res = await this.http.get<{ messages: MessageInfo[] }>('/v1/messages/inbox', queryParams);
    return res.messages ?? [];
  }

  /**
   * Acknowledge a received message.
   */
  async ack(messageId: string): Promise<boolean> {
    const res = await this.http.request<{ acknowledged: boolean }>(
      'POST',
      '/v1/messages/ack',
      undefined,
      { message_id: messageId }
    );
    return res.acknowledged ?? false;
  }
}
